### PR TITLE
ROSPlan testing

### PR DESCRIPTION
**Pull request by martinkoling**
This PR is my contribution in the scope of my Final Year Project. Please contact me if there are any questions. martin.koling@kcl.ac.uk

### CODE REVIEW

**Comment by m312z** [comment]
Hello Martin,

Looks like Travis ran the tests, and most passed. One failed, but there was an error before-hand, which looked like this:

Traceback (most recent call last):
  File "/opt/ros/kinetic/share/rostest/cmake/../../../bin/rostest", line 36, in <module>
    rostestmain()
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rostest/__init__.py", line 274, in rostestmain
    _main()
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rostest/rostest_main.py", line 159, in rostestmain
    testCase = rostest.runner.createUnitTest(pkg, test_file, options.reuse_master, options.clear, options.results_base_dir)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rostest/runner.py", line 229, in createUnitTest
    config = roslaunch.parent.load_config_default([test_file], None)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/roslaunch/config.py", line 453, in load_config_default
    raise RLException(e)
roslaunch.core.RLException: Invalid <arg> tag: rosplan_demos
ROS path [0]=/opt/ros/kinetic/share/ros
ROS path [1]=/root/catkin_ws/src/ROSPlan/rosplan_planning_system
ROS path [2]=/root/catkin_ws/install/share
ROS path [3]=/opt/ros/kinetic/share. 
Arg xml is <arg name="domain_path" value="$(find rosplan_demos)/common/domain_turtlebot.pddl"/>
[ROSUNIT] Outputting test results to /root/catkin_ws/build/rosplan_planning_system/test_results/rosplan_planning_system/rostest-test_launch_problem_interface_tests.xml
-- run_tests.py: verify result "/root/catkin_ws/build/rosplan_planning_system/test_results/rosplan_planning_system/rostest-test_launch_plan_dispatch_tests.xml"
Cannot find results, writing failure results to '/root/catkin_ws/build/rosplan_planning_system/test_results/rosplan_planning_system/MISSING-rostest-test_launch_plan_dispatch_tests.xml'
make[3]: Leaving directory '/root/catkin_ws/build/rosplan_planning_system'
[100%] Built target _run_tests_rosplan_planning_system_rostest_test_launch_plan_dispatch_tests.test
[ INFO] [1554783809.817441563]: KCL: (/rosplan_knowledge_base) Starting a PDDL Knowledge Base
[ INFO] [1554783809.837735365]: KCL: (/rosplan_planner_interface) Ready to receive
[ INFO] [1554783810.005943977]: KCL: (/rosplan_knowledge_base) Parsing domain
[ERROR] [1554783810.006047164]: KCL: (/rosplan_knowledge_base) Domain file does not exist.

**Comment by m312z** [comment]
Looks like the test has rosplan_planning_system in the package path (it is the package being tested) but not rosplan_demos, and so the domain file cannot be found by the test launch. It would be better to keep the package self-contained, and have the domain file used be in the package test folder with the others.

File: rosplan_knowledge_base/test/KnowledgeBaseTests.cpp [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+
+    ASSERT_EQ(expected_results, actual_results);
+
+    ASSERT_EQ("Jonas", srv.response.false_knowledge[0].instance_name);
+
+}
+
+int main(int argc, char **argv) {
+
+    ::testing::InitGoogleTest(&argc, argv);
```
Comment (oscar-lima) [inline rosplan_knowledge_base/test/KnowledgeBaseTests.cpp]:
```suggestion
    testing::InitGoogleTest(&argc, argv);
```

File: rosplan_knowledge_base/test/pddl/amazon-domain.pddl [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+(:predicates 
+    (at ?item - locatable ?loc - place)
+    (in ?item - locatable ?t - truck)
+
+    (link ?x ?y - place)
+    (path ?x ?y - place)		
+)
+
+
+(:action LOAD-TRUCK
```
Comment (oscar-lima) [inline rosplan_knowledge_base/test/pddl/amazon-domain.pddl]: lower case all PDDL actions

File: rosplan_planning_system/CMakeLists.txt [hunk 3]
```diff
@@ -12,6 +12,7 @@ find_package(catkin REQUIRED COMPONENTS
   rosplan_dispatch_msgs
   rosplan_knowledge_msgs
   rosplan_dependencies
+  roslib
```
Comment (oscar-lima) [inline rosplan_planning_system/CMakeLists.txt]:
```suggestion
```

File: rosplan_planning_system/CMakeLists.txt [hunk 4]
```diff
@@ -148,3 +149,28 @@ install(DIRECTORY include/rosplan_action_interface
...
+
+
+#############
+##  Tests  ##
+#############
+
+
+if (CATKIN_ENABLE_TESTING)
+    find_package(rostest REQUIRED)
+    find_package(roslib REQUIRED)
```
Comment (oscar-lima) [inline rosplan_planning_system/CMakeLists.txt]:
```suggestion
```

File: rosplan_planning_system/CMakeLists.txt [hunk 5]
```diff
@@ -148,3 +149,28 @@ install(DIRECTORY include/rosplan_action_interface
...
+    # Planner Interface
+    add_rostest_gtest(PlannerInterfaceTests test/launch/planner_interface_tests.test test/src/PlannerInterfaceTests.cpp)
+    target_link_libraries(PlannerInterfaceTests ${catkin_LIBRARIES})
```
Comment (oscar-lima) [inline rosplan_planning_system/CMakeLists.txt]: in general if you notice you are linking to catkin_LIBRARIES and not to rostest_LIBRARIES so the instruction find_package(rostest REQUIRED) , is nice because it ensures user has this package but then later on is not used. It is cleaner if you add rostest as a catkin required pkg, under : ```(catkin REQUIRED COMPONENTS``` , which is good because actually rostest is a ros pkg (you can roscd into it)

File: rosplan_planning_system/package.xml [hunk 6]
```diff
@@ -21,6 +20,8 @@
   <build_depend>rosplan_knowledge_msgs</build_depend>
   <build_depend>mongodb_store</build_depend>
   <build_depend>rosplan_dependencies</build_depend>
+  <build_depend>rostest</build_depend>
+  <build_depend>roslib</build_depend>
```
Comment (oscar-lima) [inline rosplan_planning_system/package.xml]:
```suggestion
```

File: rosplan_planning_system/package.xml [hunk 7]
```diff
@@ -32,7 +33,9 @@
   <run_depend>glut</run_depend>
   <run_depend>rosplan_dispatch_msgs</run_depend>
   <run_depend>rosplan_knowledge_msgs</run_depend>
-	<run_depend>mongodb_store</run_depend>
+  <run_depend>mongodb_store</run_depend>
+  <run_depend>rostest</run_depend>
+  <run_depend>roslib</run_depend>
```
Comment (oscar-lima) [inline rosplan_planning_system/package.xml]:
```suggestion
```

File: rosplan_planning_system/src/ProblemGeneration/ProblemInterface.cpp [hunk 8]
```diff
@@ -101,7 +101,8 @@ namespace KCL_rosplan {
 			std::ifstream problemIn(req.problem_path.c_str());
 			if(problemIn) res.problem_string = std::string(std::istreambuf_iterator<char>(problemIn), std::istreambuf_iterator<char>());
 		}
-		return success;
+		res.problem_generated = success;
+		return true;
```
Comment (oscar-lima) [inline rosplan_planning_system/src/ProblemGeneration/ProblemInterface.cpp]: this would need to go in a separate PR, is not part of the tests but is nice that as part of this PR many bugs are comming out!

File: rosplan_planning_system/test/.gitignore [hunk 9]
```diff
@@ -0,0 +1,6 @@
+.idea
```
Comment (oscar-lima) [inline rosplan_planning_system/test/.gitignore]: .idea ? mm this does not look right...

File: rosplan_planning_system/test/.gitignore [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+
+/rosplan_planning_system/test/pddl/generated_test_problem.pddl
+/rosplan_planning_system/test/pddl/plan.pddl
```
Comment (oscar-lima) [inline rosplan_planning_system/test/.gitignore]: good!

File: rosplan_planning_system/test/.gitignore [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+
+/rosplan_planning_system/__init__.py
```
Comment (oscar-lima) [inline rosplan_planning_system/test/.gitignore]: why is this needed? it looks wrong...

File: rosplan_knowledge_base/src/KnowledgeBase.cpp [hunk 12]
```diff
@@ -612,12 +612,12 @@ int main(int argc, char **argv) {
...
 	kb->use_unknowns = useUnknowns;
-
+/*
 	// wait for and clear mongoDB
 	ROS_INFO("KCL: (%s) Waiting for MongoDB", ros::this_node::getName().c_str());
 	ros::service::waitForService("/message_store/delete",-1);
 	system("mongo message_store --eval \"printjson(db.message_store.remove())\"");
-
+*/
 	ROS_INFO("KCL: (%s) Ready to receive", ros::this_node::getName().c_str());
```
[the remark this answers is not in the document]
Comment (m312z) [reply rosplan_knowledge_base/src/KnowledgeBase.cpp]: I asked for him to include this, since it is necessary for the tests to run, and we needed to do it in any case.

File: rosplan_knowledge_base/test/pddl/amazon-domain.pddl (hunk 2 above)
[the remark this answers is not in the document]
Comment (m312z) [reply rosplan_knowledge_base/test/pddl/amazon-domain.pddl]: Better yet would be to test that ROSPlan is correctly matching different cases. I know of at least one planner that converts everything to uppercase!

File: rosplan_knowledge_base/src/KnowledgeBase.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (oscar-lima) [reply rosplan_knowledge_base/src/KnowledgeBase.cpp]: Ok. This will be removed in upcoming PR from me, still in the oven but is mostly ready and should be coming soon (this week).

**Comment by m312z** [comment]
@oscar-lima @martinkoling 
I just assigned @gerardcanal to review this PR, so that Oscar can switch instead to working on it.

**Comment by oscar-lima** [comment]
closing this PR, in favor of this one: https://github.com/KCL-Planning/ROSPlan/pull/200
